### PR TITLE
Parse uevent and convert it to typed data with some more refactorings

### PR DESCRIPTION
**Pull request by Praveenrajmani**

### CODE REVIEW

File: pkg/sys/drive_discovery_linux.go [hunk 1]
```diff
@@ -738,43 +672,13 @@ func probeDevices() (devices map[string]*Device, err error) {
...
-
-	if device, err = newDevice(event, name, major, minor, strings.Contains(event["DEVPATH"], "/virtual/")); err != nil {
+func createDevice(udevData *udev.Data) (device *Device, err error) {
+	if device, err = newDevice(udevData, strings.Contains(udevData.Path, "/virtual/")); err != nil {
 		return nil, err
 	}
 
-	if event["ACTION"] == uevent.Remove {
-		return device, nil
+	if device.Name == "" {
```
Comment (balamurugana) [inline pkg/sys/drive_discovery_linux.go]: Move this error check at beginning of the function i.e. we fail early before `newDevice()`

File: pkg/sys/drive_discovery_linux.go [hunk 2]
```diff
@@ -738,43 +672,13 @@ func probeDevices() (devices map[string]*Device, err error) {
...
+		return nil, fmt.Errorf("udevData does not have valid DEVPATH %v", udevData.Path)
 	}
 
 	if device.Removable, err = getRemovable(device.Name); err != nil {
```
Comment (balamurugana) [inline pkg/sys/drive_discovery_linux.go:682]: We shouldn't attempt to populate below information for `Remove` event for obvious reason.

File: pkg/udev/udev_linux.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		PTUUID:       eventMap["ID_PART_TABLE_UUID"],
+		PTType:       eventMap["ID_PART_TABLE_TYPE"],
+		PartUUID:     eventMap["ID_PART_ENTRY_UUID"],
+		UeventFSUUID: eventMap["ID_FS_UUID"],
+		FSType:       eventMap["ID_FS_TYPE"],
+		FSUUID:       eventMap["ID_FS_UUID"],
+	}, nil
+}
+
+func NormalizeUUID(uuid string) string {
```
Comment (balamurugana) [inline pkg/udev/udev_linux.go]: Move this function to `utils` so that both `udev` and `sys` packages use this.

File: pkg/uevent/listener.go [hunk 4]
```diff
@@ -83,47 +91,172 @@ func Run(ctx context.Context, handler DeviceUEventHandler) error {
...
+		return &UEvent{
+			action:   action,
+			UDevData: udevData,
+		}, nil
+	default:
+		return nil, fmt.Errorf("invalid device action: %s", action)
+	}
 }
 
 func (l *listener) msgPeek() (int, *[]byte, error) {
```
Comment (balamurugana) [inline pkg/uevent/listener.go]: Trim the buffer to available data length and remove pointer to slice variable.

File: pkg/uevent/listener.go [hunk 5]
```diff
@@ -161,11 +294,11 @@ func (l *listener) msgRead(buf *[]byte) error {
 
 // ReadMsg allow to read an entire uevent msg
 func (l *listener) ReadMsg() ([]byte, error) {
-	_, buf, err := c.msgPeek()
+	_, buf, err := l.msgPeek()
 	if err != nil {
 		return nil, err
 	}
-	if err = c.msgRead(buf); err != nil {
+	if err = l.msgRead(buf); err != nil {
```
Comment (balamurugana) [inline pkg/uevent/listener.go]: In `msgRead()` error out after `syscall.Recvfrom()` if `n != len(buf)`. This never happens though.

File: pkg/sys/drive_discovery_linux.go (hunk 2 above)
Comment (Praveenrajmani) [reply pkg/sys/drive_discovery_linux.go:682]: yeah, correct. will fix this.

File: pkg/uevent/listener.go [hunk 6]
```diff
@@ -83,50 +92,187 @@ func Run(ctx context.Context, handler DeviceUEventHandler) error {
...
+		return l.handler.Change(ctx, device)
+	case Remove:
+		return l.handler.Remove(ctx, device)
+	default:
+		return fmt.Errorf("invalid device action: %s", uEvent.action)
+	}
+}
+
+func (l *listener) processEvents(ctx context.Context) {
+	go func() {
```
Comment (wlan0) [inline pkg/uevent/listener.go]: take threadiness as an argument and run it with those many threads

File: pkg/uevent/listener.go [hunk 7]
```diff
@@ -83,50 +92,187 @@ func Run(ctx context.Context, handler DeviceUEventHandler) error {
...
+		for {
+			select {
+			case <-ctx.Done():
+				klog.ErrorS(ctx.Err(), "context canceled")
```
Comment (wlan0) [inline pkg/uevent/listener.go]: you can just do `klog.Error(ctx.Err())`

File: pkg/uevent/listener.go [hunk 8]
```diff
@@ -83,50 +92,187 @@ func Run(ctx context.Context, handler DeviceUEventHandler) error {
...
+				return
+			default:
+				uEvent := listener.queue.pop()
+				if err := l.handle(ctx, uEvent); err != nil {
+					klog.ErrorS(err, "failed to handle an event", uEvent.action)
```
Comment (wlan0) [inline pkg/uevent/listener.go]: push it back to the end of the queue

File: pkg/uevent/listener.go [hunk 9]
```diff
@@ -23,48 +23,75 @@ import (
...
+	"github.com/minio/directpv/pkg/sys"
 	"k8s.io/klog/v2"
 )
 
 const (
 	libudev      = "libudev\x00"
 	libudevMagic = 0xfeedcafe
 	minMsgLen    = 40
+
+	Add    = "add"
```
Comment (wlan0) [inline pkg/uevent/listener.go]: use typed strings for this

type action string

const (
  ...
)

File: pkg/sys/udev_linux.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		Vendor:       eventMap["ID_VENDOR"],
+		DMName:       eventMap["DM_NAME"],
+		DMUUID:       eventMap["DM_UUID"],
+		MDUUID:       normalizeUUID(eventMap["MD_UUID"]),
+		PTUUID:       eventMap["ID_PART_TABLE_UUID"],
+		PTType:       eventMap["ID_PART_TABLE_TYPE"],
+		PartUUID:     eventMap["ID_PART_ENTRY_UUID"],
+		UeventFSUUID: eventMap["ID_FS_UUID"],
+		FSType:       eventMap["ID_FS_TYPE"],
+		FSUUID:       eventMap["ID_FS_UUID"],
```
Comment (wlan0) [inline pkg/sys/udev_linux.go]: This should be left empty

**Pull request merged**